Show Recurring Payments preview on frontend for site admins/editors

stacimc opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->


#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
If there is no Stripe account connected, but the current user is ~~an admin~~ able to edit the post, the Recurring Payments block will be displayed on the frontend anyway with the button disconnected, so that the author can preview what the block looks like on the frontend.

This is done to enable displaying a preview of the Premium Content block on the frontend for site admins/editors when the block would otherwise be hidden (due to missing plan upgrade or Stripe connection).

The plan upgrade and Stripe connection nudges should also display in the frontend _for the Premium Content block_ when relevant; the Stripe nudge will display without the `Connect` button since the url cannot be fully built in the SSR-rendered component. The user must return to the editor to connect to Stripe. The nudges do not display for the Payments block itself.

<!-- #### Jetpack product discussion -->
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

##### WPCOM site
Sync the changes to your sandbox and create a post with a Premium Content block.

For a site with a free plan:
- [ ]  Verify that you see the upgrade banner in the editor
- [ ]  Publish the post and verify that you see the block on the frontend, and that the button is disconnected (when logged in as the site administrator). The upgrade nudge should also display.
- [ ] View the post as a logged-out user and verify that you cannot see the button

For a site with a premium plan upgrade:
- [ ]  Verify that you do not see the upgrade banner in the editor, but you do see the Stripe connection banner
- [ ]  Publish the post and verify that you see the block on the frontend, and that the button is disconnected (when logged in as the site administrator). The stripe nudge should display, with no Connect button.
- [ ] View the post as a logged-out user and verify that you cannot see the button

For a site with a premium plan upgrade and Stripe connected:
- [ ]  Verify that you do not see and upgrade/Stripe warnings or nudges in the editor
- [ ]  Configure a plan for the block and publish the post. Verify that you see the block on the frontend, and that the button is connected (when logged in as the site administrator). No nudges should display.
- [ ] View the post as a logged-out user and verify that you can see the button

Repeat tests with a Jetpack site.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
*

Commit:
Show Recurring Payments preview on frontend for admin users

If there is no Stripe account connected, but the current user is
an admin, display the block on the frontend anyway but disconnect
the button.

apeatling commented:
**On WordPress.com**

For a site with a free plan:
- [x]  Verify that you see the upgrade banner in the editor
- [x]  Publish the post and verify that you see the block on the frontend, and that the button is disconnected (when logged in as the site administrator)
- [x] View the post as a logged-out user and verify that you cannot see the button

For a site with a premium plan upgrade:
- [x]  Verify that you do not see the upgrade banner in the editor, but you do see the Stripe connection banner
- [x]  Publish the post and verify that you see the block on the frontend, and that the button is disconnected (when logged in as the site administrator)
- [x] View the post as a logged-out user and verify that you cannot see the button

For a site with a premium plan upgrade and Stripe connected:
- [x]  Verify that you do not see and upgrade/Stripe warnings or nudges in the editor
- [ ]  **Configure a plan for the block and publish the post. Verify that you see the block on the frontend, and that the button is connected (when logged in as the site administrator)**
- [x] View the post as a logged-out user and verify that you can see the button

Everything worked on dotcom with the exception of being able to see anything on the front end of a premium plan site as am admin, once I had connected Stripe. It did however show up as a regular visitor to the site.

stacimc commented:
> Everything worked on dotcom with the exception of being able to see anything on the front end of a premium plan site as an admin, once I had connected Stripe.

@apeatling Were you testing with the Payments block or Premium Content?

With the Premium Content block, if the block is configured properly (premium plan, Stripe connected), it will automatically display the Subscriber view on the frontend if the user is logged in and has access -- and an admin user is considered to have access. If you were testing with Premium Content, is it possible you hadn't added anything to the Subscriber view yet and were seeing that (empty) view?

apeatling commented:
>  is it possible you hadn't added anything to the Subscriber view yet and were seeing that (empty) view?

Just checked and confirmed this was the issue. Opened an issue to add default content: https://github.com/Automattic/view-design/issues/138

Commit:
Server side render stripe nudge on frontend

Commit:
Pass isPremiumChildContent to render callback

stacimc commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -392,7 +431,8 @@ public function register_gutenberg_block() {
 				array(
-					'render_callback' => array( $this, 'render_button' ),
+					'render_callback' =>  array( $this, 'render_button' ),
+					'uses_context' => array( 'isPremiumContentChild' ),
```
The goal here is to make `$block->context` available in the render callback. This is possible when using `register_block_from_metadata` to register the block (see Social Link block as an example).

I stepped through the implementation of `register_block_type_from_metadata`, and it has the effect of passing this `uses_context` setting through to `register_block_type`. I'm attempting here to hard-code this for Recurring Payments, since I've had difficulty restructuring this block to use a `block.json` like Social Link and getting the correct path for that file once it is created.

Commit:
Use isPremiumContentChild context in block registration

Commit:
Refactor StripeNudge to server-side-render simple component

Much like the upgrade nudge, this refactors the Stripe Nudge so
that we can take just the simple component (without any props
computation) to be used for server side rendering.

Commit:
Conditionally render the Payments preview

Render a preview of the Payments button when the user is an admin,
the block is part of a Premium Content block, and the Premium
Content block is not properly configured (either because the site
requires an upgrade or a Stripe connection)

Commit:
Remove Connect button from server side rendered Stripe nudge

The Connect url cannot be built fully on the backend, so for now
we are removing the Connect button from the Stripe nudge when it
is server side rendered; the user must return to the editor to
connect.

jeherve reviewed (commented):
If you run `yarn php:lint modules/memberships/class-jetpack-memberships.php`, you should get some indicators about issues that you should be able to fix  to avoid any PHPCS issues.

Commit:
Clean up doc strings and linting.

Commit:
Revert whitespace changes to components.php

apeatling commented:
Rebased

apeatling commented:
I've retested this in the WordPress.com environment and can confirm this is all working correctly:

* ✅ On a free plan, I can edit the block and publish. I see an upgrade banner in the editor. I see a notice on the front end as an admin telling me that I need to purchase a personal plan for this block to work. As a visitor to the site I cannot see any content. 
* ✅ On a premium plan I can edit the block and publish. I do not see an upgrade banner. With Stripe connected I'm able to see the subscriber content in the front end with no notice as an admin. As a visitor I can see the visitor content rendered correctly. With Stripe disconnected I see a notice in the front end asking me to connect Stripe before the block will work, as a visitor I see nothing.

apeatling commented:
Also confirming that this is working correctly for me on a local Jetpack install with the editing toolkit plugin installed. I saw the correct notices and access on the front end when using a site with a paid plan, and connected/disconnected Stripe. ✅

apeatling reviewed (commented):
Confirmed this is working in on WordPress.com with/without plan, and on a Jetpack site with paid plan and editing toolkit available. Looks good for a final Crew review. 👍

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -389,10 +459,13 @@ public function register_gutenberg_block() {
 		}
 
 		if ( self::is_enabled_jetpack_recurring_payments() ) {
+			$deprecated = function_exists( 'gutenberg_get_post_from_context' );
```
How far back does this go? Jetpack now requires WordPress 5.5, so we may not need that check anymore. I've tried to figure that out by `blame`ing `lib/compat.php` in Gutenberg, but I may not be looking in the right place.

stacimc replied:
I tried searching for this function in the source for WP5.5 and could not find it, so I think this might still be necessary.

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -357,6 +417,17 @@ public static function get_connected_account_id() {
+	public static function user_can_edit() {
+		$user = wp_get_current_user();
+		// phpcs:ignore ImportDetection.Imports.RequireImports.Symbol
+		return 0 !== $user->ID && current_user_can( 'edit_post', get_the_ID() );
```
If we want to check for admins, would checking for `manage_options` be enough here? I worry that with this approach, we may target authors as well?

stacimc replied:
I believe we do want to target authors, since the intention is for a user editing a post to be able to preview their change on the frontend even if it is not fully configured. My use of 'admin' is confusing; I'll make sure to change that.

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -227,21 +227,81 @@ public function allow_sync_post_meta( $post_meta ) {
+		$requires_stripe_connection = ! $this->get_connected_account_id();
+
+		$requires_upgrade = false;
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM && function_exists( 'has_any_blog_stickers' ) ) {
+			$site_id          = get_current_blog_id();
+			$requires_upgrade = has_any_blog_stickers( array( 'personal-plan', 'premium-plan', 'business-plan', 'ecommerce-plan' ), $site_id );
```
We do things a bit differently elsewhere? 
https://github.com/Automattic/jetpack/blob/9222df3c888953e6f0d867e4c4d0e0e47c1ad5f2/class.jetpack-gutenberg.php#L1017-L1021

Should we / Could we take a similar approach here?

We also have `is_enabled_jetpack_recurring_payments` in this file. Maybe that's an opportunity to merge things a bit?

stacimc replied:
I'm not very familiar with this code, but when I inspect `$features_data` on a site which I know supports recurring payments (WPCOM site with a premium plan), I don't see any entry for recurring-payments.

`is_enabled_jetpack_recurring_payments` does almost what we want but not quite -- rather than checking whether the block is supported by the site's plan, it checks whether it ought to be enabled. This is true for Jetpack sites when the upgrade nudge is enabled, even if the block is not supported.

I agree that's a good opportunity to merge things a bit; I'll try to refactor that function slightly.

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -227,21 +227,81 @@ public function allow_sync_post_meta( $post_meta ) {
+			$site_id          = get_current_blog_id();
+			$requires_upgrade = has_any_blog_stickers( array( 'personal-plan', 'premium-plan', 'business-plan', 'ecommerce-plan' ), $site_id );
+		} else {
+			$requires_upgrade = Jetpack::is_active() && Jetpack_Plan::supports( 'recurring-payments' );
```
Do we need to check for the `jetpack_block_editor_enable_upgrade_nudge` filter here, like we do elsewhere?

stacimc replied:
No, we need to exclude that check here -- which is also unfortunately why we can't just use `is_enabled_jetpack_recurring_payments` for the check.

For the purpose of displaying the block preview, we need to check whether the site requires an upgrade in order to fully support the block. The filter indicates that the block should be enabled with the upgrade nudge if is not supported. It is used to determine whether the block should be registered.

Commit:
Clarify that the preview is shown to users with editing permission

Commit:
Refactor block support check to allow code reuse

jeherve commented on modules/memberships/class-jetpack-memberships.php:
```diff
@@ -358,21 +412,46 @@ public static function get_connected_account_id() {
+	 */
+	public static function user_can_edit() {
+		$user = wp_get_current_user();
+		// phpcs:ignore ImportDetection.Imports.RequireImports.Symbol
```
Do we need this here?

```suggestion
```

jeherve reviewed (dismissed):
This looks good to me. I only have a tiny remark, but that's not a blocker, it can be addressed in a follow-up.

Edit: hm, it looks like you'll need to rebase to fix those unrelated tests failures. May be an opportunity to deal with my suggestion :)

stacimc commented:
Rebased

stacimc commented:
Rebased for the dashboard, extensions, and search build step.

jeherve commented:
@stacimc I ran into issues when testing D52683-code on WordPress.com simple. I left a comment there. Do you think you could take a look?

Thank you!

jeherve commented:
r220422-wpcom
